pyquil/api/_engagement_manager.py

```diff
@@ -0,0 +1,76 @@
+##############################################################################
+# Copyright 2016-2021 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+from datetime import datetime
+from typing import Dict, Optional
+
+import httpx
+from dateutil.parser import parse as parsedate
+from dateutil.tz import tzutc
+from qcs_api_client.client import QCSClientConfiguration
+from qcs_api_client.models import EngagementWithCredentials, CreateEngagementRequest
+from qcs_api_client.operations.sync import create_engagement
+
+from pyquil.api._qcs_client import qcs_client
+
+
+class EngagementManager:
+    """
+    Fetches (and caches) engagements for use when accessing a QPU.
+    """
+
+    def __init__(self, *, client_configuration: QCSClientConfiguration) -> None:
+        """
+        Instantiate a new engagement manager.
+
+        :param client_configuration: Client configuration, used for refreshing engagements.
+        """
+        self._client_configuration = client_configuration
+        self._cached_engagements: Dict[str, EngagementWithCredentials] = {}
```

Comment:
I believe get/set ops on a dictionary are thread-safe, so this shouldn't cause issues when running parallel quantum programs using the same engagement manager